File: atcoder/abc202_e.py
# https://atcoder.jp/contests/abc202/tasks/abc202_e

import sys
import bisect
input = sys.stdin.buffer.readline
sys.setrecursionlimit(10 ** 7)

def main():
    N = int(input())
    P = list(map(int, (input().split())))
    G = [[] for _ in range(N)]
    depth   = [-1] * N
    in_num  = [-1] * N
    out_num = [-1] * N
    k = 0
    ls = [[] for _ in range(N)]
    for i in range(N-1):
        G[P[i]-1].append(i+1)
        # G[i+1].append(P[i]-1)

    def dfs(now, d, G, depth, k, in_num, out_num):
        in_num[now] = k
        k += 1
        depth[now] = d
        for next in G[now]:
            # if depth[next]!=-1: continue
            k = dfs(next, d+1, G, depth, k, in_num, out_num)
        out_num[now] = k
        return k

    dfs(0, 0, G, depth, k, in_num, out_num)
    for i in range(N):
        ls[depth[i]].append(in_num[i])
        # print(i, depth[i], in_num[i], out_num[i])
    for i in range(N):
        ls[i].sort()
        # print(i, ls[i])

    Q = int(input())
    for qi in range(Q):
        U, D = map(int, input().split())
        U -= 1
        ans = bisect.bisect_left(ls[D], out_num[U]) - bisect.bisect_left(ls[D], in_num[U])
        print(ans)
    return

main()


# def main():
#     N = int(input())
#     P = list(map(int, (input().split())))
#     G = [[] for _ in range(N)]
#     depth = [-1 for _ in range(N)]
#     child = [set() for _ in range(N)]
#     for i in range(N-1):
#         G[P[i]-1].append(i+1)
#         G[i+1].append(P[i]-1)

#     def dfs(now, d, G, depth, child):
#         depth[now] = d
#         child_set = set([now])
#         for next in G[now]:
#             if depth[next]!=-1: continue
#             child_set |= dfs(next, d+1, G, depth, child)
#         child[now] |= child_set
#         return child_set

#     child[0] |= dfs(0, 0, G, depth, child)
#     # for i in range(N):
#     #     print(i, depth[i], child[i])

#     Q = int(input())
#     for qi in range(Q):
#         U, D = map(int, input().split())
#         ans = 0
#         for i in child[U-1]:
#             if depth[i] == D:
#                 ans += 1
#         print(ans)
#     return

# main()

File: atcoder/abc202_e.cpp
// https://atcoder.jp/contests/abc202/tasks/abc202_e
#include<iostream>
// #include<algorithm>
// #include<string>
// #include<numeric>
// #include<vector>
// #include<map>
// #include<tuple>
// #include<set>
// #include<queue>
// #include<deque>
// #include<regex>
// #include<bitset>
// #include<iomanip>
// #include<complex>
// #include<stack>
// #include<functional>

#include<bits/stdc++.h>
using namespace std;
// #include <atcoder/all>
// using namespace atcoder;
#define rep(i,n) for (int i = 0; i < (n); ++i)
#define drep(i,n) for(int i = (n-1); i >= 0; i--)
#define all(v) (v).begin(),(v).end()
#define maxs(x,y) (x = max(x,y))
#define mins(x,y) (x = min(x,y))
template <class T> bool chmax(T &a, const T &b) { if (a < b) { a = b; return 1; } return 0; }
template <class T> bool chmin(T &a, const T &b) { if (b < a) { a = b; return 1; } return 0; }
template <class T> T gcd(T a, T b) { return b ? gcd(b, a % b) : a; }
template <class T> T lcm(T a, T b) { return a/gcd(a,b)*b; }
typedef pair<int, int> P;
typedef long long ll;
const int INF = 1001001001;
const ll LINF = 1001002003004005006ll;
const ll MOD = 1e9+7;

const int MAX_V = 200005;
vector<int> to[MAX_V];
int dep[MAX_V];    // each node depth
int in[MAX_V];     // each node in order (Euler Tour)
int out[MAX_V];    // each node out order (Euler Tour)
int k;             // order number (Euler Tour)
vector<int> ls[MAX_V];   // each depth in order list

void dfs(int v, int d) {
    in[v] = k; k++;    // in order number (Euler Tour)
    dep[v] = d;        // depth
    for (int u: to[v]) dfs(u, d+1);
    out[v] = k;        // out order number (Euler Tour)
    return;
}

int main() {
    cin.tie(nullptr);
    ios::sync_with_stdio(false);
	int n;
	cin >> n;
    rep(i,n-1) {
        int p;
        cin >> p;
        to[p-1].push_back(i+1);
    }
    dfs(0, 0);
    rep(i,n) ls[dep[i]].push_back(in[i]);       // each depth in order list
    rep(i,n) sort(ls[i].begin(), ls[i].end());  // sort for binary search

    int q;
    cin >> q;

    auto f = [&](int d, int r) {  // binary search
        return int(lower_bound(ls[d].begin(), ls[d].end(), r) - ls[d].begin());
    };
    rep(qi, q) {
        int u, d;
        cin >> u >> d;
        --u;
        int ans = f(d, out[u]) - f(d, in[u]);  // count nodes by binary search
        printf("%d\n", ans);
    }

	return 0;
}
